File: NeetCode/backtracking/permutations.cpp
#include <vector>

using namespace std;

class Solution {
  vector<vector<int>> ans;

 public:
  vector<vector<int>> permute(vector<int>& nums) {
    vector<int> p;
    dfs(nums, 0, p);
    return ans;
  }

  void dfs(vector<int>& nums, int index, vector<int>& p) {
    if (index == nums.size()) {
      ans.push_back(p);
      return;
    }
    for (int i = index; i < nums.size(); i++) {
      p.push_back(nums[i]);
      swap(nums, i, index);
      dfs(nums, index + 1, p);
      swap(nums, i, index);
      p.pop_back();
    }
  }

 private:
  void swap(vector<int>& nums, int i, int j) {
    int temp = nums[i];
    nums[i] = nums[j];
    nums[j] = temp;
  }
};

File: NeetCode/backtracking/permutations.java
class Solution {
  public List<List<Integer>> permute(int[] nums) {
    List<List<Integer>> ans = new ArrayList<>();
    dfs(nums, 0, new ArrayList<>(), ans);
    return ans;
  }

  private void dfs(int[] nums, int index, List<Integer> p, List<List<Integer>> ans) {
    if (index == nums.length) {
      ans.add(new ArrayList<>(p));
      return;
    }
    for (int i = index; i < nums.length; i++) {
      p.add(nums[i]);
      swap(nums, index, i);
      dfs(nums, index + 1, p, ans);
      p.remove(p.size() - 1);
      swap(nums, index, i);
    }
  }

  private void swap(int[] nums, int i, int j) {
    int temp = nums[i];
    nums[i] = nums[j];
    nums[j] = temp;
  }
}

// TC: O(n! * n)
// SC: O(n! * n)
